src/main/java/software/amazon/documentdb/jdbc/sshtunnel/DocumentDbSshTunnelServer.java

```diff
@@ -115,69 +107,294 @@ private DocumentDbSshTunnelServer(final DocumentDbSshTunnelServerBuilder builder
+        } catch (Exception e) {
+            throw logException(e);
+        }
+    }
+
+    private static SshPortForwardingSession getPortForwardingSession(
+            final DocumentDbConnectionProperties connectionProperties,
+            final Session session) throws JSchException {
+        final Pair<String, Integer> clusterHostAndPort = getHostAndPort(
+                connectionProperties.getHostname(), DEFAULT_DOCUMENTDB_PORT);
+        final int localPort = session.setPortForwardingL(
+                LOCALHOST, 0, clusterHostAndPort.getLeft(), clusterHostAndPort.getRight());
+        return new SshPortForwardingSession(session, localPort);
+    }
+
+    private static Pair<String, Integer> getHostAndPort(
+            final String hostname,
+            final int defaultPort) {
+        final String clusterHost;
+        final int clusterPort;
+        final int portSeparatorIndex = hostname.indexOf(':');
+        if (portSeparatorIndex >= 0) {
+            clusterHost = hostname.substring(0, portSeparatorIndex);
+            clusterPort = Integer.parseInt(
+                    hostname.substring(portSeparatorIndex + 1));
+        } else {
+            clusterHost = hostname;
+            clusterPort = defaultPort;
+        }
+        return new ImmutablePair<>(clusterHost, clusterPort);
+    }
+
+    private static void connectSession(
+            final DocumentDbConnectionProperties connectionProperties,
+            final JSch jSch,
+            final Session session) throws SQLException {
+        setSecurityConfig(connectionProperties, jSch, session);
+        try {
+            session.connect();
+        } catch (JSchException e) {
+            throw logException(e);
+        }
+    }
+
+    private static void addIdentity(
+            final DocumentDbConnectionProperties connectionProperties,
+            final JSch jSch) throws JSchException {
+        final String privateKeyFileName = getPath(connectionProperties.getSshPrivateKeyFile(),
+                DocumentDbConnectionProperties.getDocumentDbSearchPaths()).toString();
+        LOGGER.debug("SSH private key file resolved to '{}'.", privateKeyFileName);
+        // If passPhrase protected, will need to provide this, too.
+        final String passPhrase = !isNullOrWhitespace(connectionProperties.getSshPrivateKeyPassphrase())
+                ? connectionProperties.getSshPrivateKeyPassphrase()
+                : null;
+        jSch.addIdentity(privateKeyFileName, passPhrase);
+    }
+
+    private static Session createSession(
+            final DocumentDbConnectionProperties connectionProperties,
+            final JSch jSch) throws SQLException {
+        final String sshUsername = connectionProperties.getSshUser();
+        final Pair<String, Integer> sshHostAndPort = getHostAndPort(
+                connectionProperties.getSshHostname(), DEFAULT_SSH_PORT);
+        setKnownHostsFile(connectionProperties, jSch);
+        try {
```

Comment:
It is possible an exception thrown from setKnownHostsFile but it is not caught here.

Reply:
Yes. You're right. However, it is not caught here and just passed to the caller.